polybuild/polybuild.py

```diff
@@ -0,0 +1,358 @@
+        ret_code = subprocess.call(command)
+        if ret_code != 0:
+            print(f"Error! Failed to execute compile command: {compile_command}")
+            return False
+        return True
+
+    def poly_opt(self, input_file: str, bitcode_file: str) -> bool:
+        opt_command = ["opt", "-O0", "-load", self.meta.compiler_dir + "../pass/libDataFlowSanitizerPass.so"]
+        ignore_list_files: Optional[List[str]] = self.poly_add_inst_lists("ignore_lists")
+        if ignore_list_files is None:
+            print("Error! Failed to add ignore lists")
+            return False
+        track_list_files: Optional[List[str]] = self.poly_add_inst_lists("track_lists")
+        if track_list_files is None:
+            print("Error! Failed to add track_lists")
+            return False
+        for file in ignore_list_files:
+            opt_command.append("-polytrack-dfsan-abilist=" + file)
+        for file in track_list_files:
+            opt_command.append("-polytrack-dfsan-abilist=" + file)
+        opt_command += [input_file, "-o", bitcode_file]
+        command = shlex.split(" ".join(shlex.quote(arg) for arg in opt_command))
+        ret_code = subprocess.call(command)
+        if ret_code != 0:
+            print("Error! opt command failed!")
+            return False
+        if not os.path.exists(bitcode_file):
+            print("Error! Bitcode file does not exist!")
+            return False
+        return True
+
+    def poly_instrument(self, input_file, output_file, bitcode_file, libs) -> bool:
+        res = self.poly_opt(input_file, bitcode_file)
+        if not res:
+            print(f"Error instrumenting bitcode {input_file} with opt!")
+            return False
+        res = self.poly_compile(bitcode_file, output_file, libs)
+        if not res:
+            print(f"Error compiling bitcode!")
+            return False
+        return True
+
+    # TODO add qUnusedArgs here
+    def poly_build(self, argv) -> bool:
+        compile_command = []
+        if self.meta.is_cxx:
+            compile_command.append("gclang++")
+        else:
+            compile_command.append("gclang")
+        compile_command += ["-pie", "-fPIC"]
+        if self.meta.is_cxx:
+            compile_command.append("-stdlib=libc++")
+            compile_command.append("-nostdinc++")
+            compile_command.append("-I" + self.meta.compiler_dir + "/../cxx_libs/include/c++/v1/")
+            compile_command.append("-L" + self.meta.compiler_dir + "/../cxx_libs/lib/")
+        for arg in argv[1:]:
+            compile_command.append(arg)
+        is_linking = self.poly_is_linking(argv)
+        if is_linking:
+            # If its cxx, link in our c++ libs
+            if self.meta.is_cxx:
+                compile_command += ["-lc++", "-lc++abipoly", "-lc++abi", "-lpthread"]
+
+        command = shlex.split(" ".join(shlex.quote(arg) for arg in compile_command))
+        res = subprocess.call(command)
```

Comment:
Same here, I don't think you need `shlex.split` if you pass `compile_command` directly to `subprocess.call`. _However!_ If you do that, then you might need to split up some of the elements of the compile command. For example, I think `compile_command.append("-I" + self.meta.compiler_dir + "/../cxx_libs/include/c++/v1/")` would have to become `compile_command += ["-I", self.meta.compiler_dir, "/../cxx_libs/include/c++/v1/"]`.

Reply:
I don't think it splits it up that way, when you do the "-I" + string + .... and include it into the list it gets treated as one arg I believe. At least thats what I saw from my command line logging output. I can check again

Reply:
Removed shlex so that should all be good.